connectors/memory/memory.go

```diff
@@ -0,0 +1,493 @@
+	// scoot over remaining elements
+	copy(entityRef[encodedPartitionKey][offset+1:], entityRef[encodedPartitionKey][offset:])
+	// and plunk value into appropriate location
+	entityRef[encodedPartitionKey][offset] = values
+	return nil
+}
+func merge(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) {
+	for k, v := range from {
+		into[k] = v
+	}
+}
+
+// Remove deletes a single row
+func (c *Connector) Remove(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.data[ei.Def.Name] == nil {
+		return &dosa.ErrNotFound{}
+	}
+	entityRef := c.data[ei.Def.Name]
+	encodedPartitionKey := partitionKeyBuilder(ei, values)
+	if entityRef[encodedPartitionKey] == nil {
+		return &dosa.ErrNotFound{}
+	}
+	partitionRef := entityRef[encodedPartitionKey]
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
+		return &dosa.ErrNotFound{}
+	}
+
+	// no clustering keys? Simple, delete this
+	if len(ei.Def.ClusteringKeySet()) == 0 {
+		entityRef[encodedPartitionKey] = nil
+		return nil
+	}
+	found, offset := findInsertionPoint(ei, partitionRef, values)
+	if found {
+		entityRef[encodedPartitionKey] = append(entityRef[encodedPartitionKey][:offset], entityRef[encodedPartitionKey][offset+1:]...)
+		return nil
+	}
+	return &dosa.ErrNotFound{}
+}
+
+// Range returns a slice of data from the datastore
+func (c *Connector) Range(_ context.Context, ei *dosa.EntityInfo, columnConditions map[string][]*dosa.Condition, fieldsToRead []string, token string, limit int) ([]map[string]dosa.FieldValue, string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if c.data[ei.Def.Name] == nil {
+		return nil, "", &dosa.ErrNotFound{}
+	}
+	entityRef := c.data[ei.Def.Name]
+
+	// find the equals conditions on each of the partition keys
+	values := map[string]dosa.FieldValue{}
+	for _, pk := range ei.Def.Key.PartitionKeys {
+		// TODO: assert len(columnConditions[pk] == 1
+		// TODO: assert columnConditions[pk][0].Op is equals
+		values[pk] = columnConditions[pk][0].Value
+	}
+
+	encodedPartitionKey := partitionKeyBuilder(ei, values)
+	partitionRef := entityRef[encodedPartitionKey]
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
+		return nil, "", &dosa.ErrNotFound{}
+	}
+	// hunt through the partitionRef and return values that match search criteria
+	// TODO: This can be done much faster using a binary search
+	startinx, endinx := 0, len(partitionRef)-1
+	for startinx < len(partitionRef) && !matchesClusteringConditions(ei, columnConditions, partitionRef[startinx]) {
+		startinx++
+	}
+	// TODO: adjust startinx with a passed in token
+	for endinx >= startinx && !matchesClusteringConditions(ei, columnConditions, partitionRef[endinx]) {
+		endinx--
+
+	}
+	if endinx <= startinx {
+		return nil, "", &dosa.ErrNotFound{}
+	}
+	// TODO: enforce limits and return a token when there are more rows
```

Comment:
No, we need to construct a token so this can be continued. The token needs to be a serialized `[]Conditions`, created based on the contents of the last row returned.

This is a tricky subject that needs a ton of documentation once it's written. Let's get this merged first and then address it.